db/db_impl/db_impl.cc

```diff
@@ -4727,7 +4728,30 @@ Status DBImpl::CreateColumnFamilyWithImport(
 }
 
 Status DBImpl::VerifyChecksum(const ReadOptions& read_options) {
+  return VerifyChecksum(read_options, /*use_file_checksum=*/false);
+}
+
+Status DBImpl::VerifyChecksum(const ReadOptions& read_options,
+                              bool use_file_checksum) {
   Status s;
+
+  std::string file_checksum_func_name(kUnknownFileChecksumFuncName);
+  if (use_file_checksum) {
+    FileChecksumGenFactory* const file_checksum_gen_factory =
+        immutable_db_options_.file_checksum_gen_factory.get();
+    if (!file_checksum_gen_factory) {
+      s = Status::InvalidArgument(
+          "use_file_checksum cannot be true if file checksum is disabled");
+      return s;
+    } else {
+      FileChecksumGenContext file_checksum_context;
+      std::unique_ptr<FileChecksumGenerator> file_checksum_gen =
+          file_checksum_gen_factory->CreateFileChecksumGenerator(
+              file_checksum_context);
+      file_checksum_func_name = file_checksum_gen->Name();
```

Comment:
This seems like an awful expensive operation to get the name of the checksum func.  Is there a cheaper alternative?

Reply:
I checked, and didn't find one. From `FileChecksumGenFactory`, the name is different from the name returned from `FileChecksumGenerator`. Let me know if I miss anything @zhichao-cao .

Reply:
Currently, no. User provides the shared_ptr of FileChecksumGenFactory and it brings the factory name. The checksum function name is brought by the generator itself. We can add the get checksum function name to the Factory, but it will break the flexibility. It is possible that there are more than one generator in a factory and only the one match the file_checksum_context will be returned.

Reply:
Is it possible that the application specified a different name for the generator in `FileChecksumContext` even if the same underlying algorithm is used?

Reply:
It is possible. We do not have the role or contract for the naming of the generator. E.g., we can call it "crc32c" or "CRC32C".

Reply:
Thanks @zhichao-cao for the info. Then we need to handle this case, too. If the name can be anything, it is a little hard.

Reply:
Just double check your question: user implement the role to get the FileChecksumGenerator based on the name passed in FileChecksumContext.

Reply:
hmm, I wonder how much effort we should put here to detect the equality of checksum function name. Should we just require that they be the same (case sensitive) so that file checksum verification can be done?

Reply:
Actually, if we consider a production case, application should always use the same FileChecksumGenFactory. So the name should be exactly the same. Only in the case that the DB is opened by a different application, then, there might be the possibility that same algorithm is used with different name.

Reply:
Looks like the file_checksum_context does not carry the name, in this case, you will get the default generator. So the file_checksum_func_name is the default generator's name. As mentioned before, in most cases, there is only one generator type in a factory. But it is possible, there are more than one generator in a factory. If not the default one is used in the file, we will have wrong verification results.

Reply:
I think the solution will be to leverage your change in #7586.